File: src/main/kotlin/dev/jefy/connectpro/user/domain/service/TokenManager.java
package dev.jefy.connectpro.user.domain.service;


import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import dev.jefy.connectpro.user.applicaion.exceptions.TokenHasExpiredException;
import dev.jefy.connectpro.user.applicaion.exceptions.TokenNotFoundException;
import dev.jefy.connectpro.user.domain.model.Token;
import dev.jefy.connectpro.user.domain.repository.TokenRepository;
import dev.jefy.connectpro.user.domain.vo.OtpCode;
import dev.jefy.connectpro.user.domain.vo.TokenId;
import dev.jefy.connectpro.user.domain.vo.UserId;
import lombok.RequiredArgsConstructor;

/**
 * @author Jôph Yamba
 */
@Service
@RequiredArgsConstructor
public class TokenManager {
    @Value("${security.verification-token.expiration-time}")
    private long verificationTokenExpirationTimeInMinutes;
    private final TokenRepository tokenRepository;
    private final OtpCodeGenerator otpCodeGenerator;

    public Token generateAndSave(UserId userId) {
        OtpCode otpCode = otpCodeGenerator.generate();
        Token token = new Token(userId, otpCode, verificationTokenExpirationTimeInMinutes);
        tokenRepository.save(token);
        return token;
    }

    public Token validateToken(TokenId tokenId, OtpCode code) {
        Token token = tokenRepository.findByIdAndCode(tokenId,code)
                .orElseThrow(TokenNotFoundException::new);
        if (token.isExpired()) {
            throw new TokenHasExpiredException();
        }
        token.validate();
        return token;
    }
    
    public void save(Token token) {
        tokenRepository.save(token);
    }

    public UserId checkValidity(TokenId tokenId) {
        Token token = tokenRepository.findById(tokenId)
                .orElseThrow(TokenNotFoundException::new);
        if (token.isExpired()) {
            throw new TokenHasExpiredException();
        }
        if (token.isNotValidated()) {
            throw new IllegalStateException("token not validated");
        }
        return  token.getUserId();
    }
}

File: src/main/kotlin/dev/jefy/connectpro/user/domain/service/TokenManager.kt
package dev.jefy.connectpro.user.domain.service

import dev.jefy.connectpro.shared.infrastructure.config.SecurityProperties
import dev.jefy.connectpro.user.application.exceptions.TokenHasExpiredException
import dev.jefy.connectpro.user.application.exceptions.TokenNotFoundException
import dev.jefy.connectpro.user.domain.model.Token
import dev.jefy.connectpro.user.domain.repository.TokenRepository
import dev.jefy.connectpro.user.domain.vo.OtpCode
import dev.jefy.connectpro.user.domain.vo.TokenId
import dev.jefy.connectpro.user.domain.vo.UserId
import org.springframework.stereotype.Service

/**
 * @author  Jôph Yamba
 */
@Service
class TokenManager(
    private val securityProperties: SecurityProperties,
    private val tokenRepository: TokenRepository,
    private val otpCodeGenerator: OtpCodeGenerator
) {

    fun generateAndSave(userId: UserId): Token {
        val otpCode = otpCodeGenerator.generate()
        val token = Token(userId, otpCode, securityProperties.otpToken.expirationTime)
        tokenRepository.save(token)
        return token
    }

    fun validateToken(tokenId: TokenId, code: OtpCode): Token {
        val token = tokenRepository.findByIdAndCode(tokenId, code)
            .orElseThrow { TokenNotFoundException() }
        if (token.isExpired()) {
            throw TokenHasExpiredException()
        }
        token.validate()
        return token
    }

    fun save(token: Token) {
        tokenRepository.save(token)
    }

    fun checkValidity(tokenId: TokenId): UserId {
        val token = tokenRepository.findById(tokenId)
            .orElseThrow { TokenNotFoundException() }
        if (token.isExpired()) {
            throw TokenHasExpiredException()
        }
        if (token.isNotValidated()) {
            throw IllegalStateException("token not validated")
        }
        return token.userId
    }
}